compiler/x/codegen/BinaryEvaluator.cpp

```diff
@@ -1937,13 +1937,59 @@ TR::Register *OMR::X86::TreeEvaluator::integerMulEvaluator(TR::Node *node, TR::C
    TR::Node     *firstChild     = node->getFirstChild();
    TR::Node     *secondChild    = node->getSecondChild();
    bool          nodeIs64Bit    = TR::TreeEvaluator::getNodeIs64Bit(node, cg);
+   bool isMemRef                = firstChild->getOpCode().isMemoryReference();
+   TR_X86OpCodes opCode, regOpCode, memOpCode, movOpCode;
    intptrj_t    value;
 
    if (node->isDualCyclic())
       {
       return TR::TreeEvaluator::integerDualMulEvaluator(node, cg);
       }
 
+   // Opcode evaluation for bmul, smul imul and lmul
+   switch (secondChild->getDataType())
+      {
+      case TR::Int8:
+         value = secondChild->getByte();
+         opCode = isMemRef ? IMUL1AccMem : IMUL1AccReg;
+         regOpCode = IMUL1AccReg;
+         memOpCode = IMUL1AccMem;
+         movOpCode = MOV1RegReg;
+         break;
+
+      case TR::Int16:
+         value = secondChild->getShortInt();
+         if (isMemRef)
+            opCode = (value >= -128 && value <= 127) ? IMUL2RegMemImms : IMUL2RegMemImms;
+         else
+            opCode = (value >= -128 && value <= 127) ? IMUL2RegRegImms : IMUL2RegRegImm2;
+
+         regOpCode = IMUL2RegReg;
+         memOpCode = IMUL2RegMem;
+         movOpCode = MOV2RegReg;
+         break;
+
+      case TR::Int32:
+      case TR::Int64:
+         if(nodeIs64Bit)
+            value = secondChild->getLongInt();
+         else
+            value = secondChild->getInt();
+
+         if (isMemRef)
+            opCode = (value >= -128 && value <= 127) ? IMULRegMemImms(nodeIs64Bit) : IMULRegMemImm4(nodeIs64Bit);
+         else
+            opCode = (value >= -128 && value <= 127) ? IMULRegRegImms(nodeIs64Bit) : IMULRegRegImms(nodeIs64Bit);
+         
+         regOpCode = IMULRegReg(nodeIs64Bit);
+         memOpCode = IMULRegMem(nodeIs64Bit);
+         movOpCode = MOVRegReg(nodeIs64Bit);
+         break;
+
+      default:
+         TR_ASSERT(0, "Unexpected data type for logical node [" POINTER_PRINTF_FORMAT "]\n", secondChild);
```

Comment:
This needs to be a TR_ASSERT_FATAL so that it will always fire since hitting the default case means we don't know what to do.